### PR TITLE
Remove Redux dependency from the wpcom localization middleware

### PR DESCRIPTION
**Pull request by jsnajdr**
This PR is a followup to #38047 that further simplifies the `lib/wp/localization` middleware for the `wpcom` library that adds a `locale` query param to every request, based on current locale.

It removes the Redux dependency by getting the current locale slug and locale variant not from Redux, but directly from the `i18n-calypso` library.

After the Redux dependency is removed, we also no longer need to bind the middleware to Redux by importing and calling `bindState` in the entrypoint boot code.

To make that work, I also had to add the `i18n.getLocaleVariant` method to `i18n-calypso`, in order to make special locales like `sr_latin` and `de_formal` behave correctly.

To make unit tests pass, I had to remove unneeded and incomplete mocking of `i18n-calypso` from several Enzyme tests. When a unit test makes a network request (intercepted by `nock`), the `i18n.getLocaleVariant()` and `i18n.getLocaleSlug()` are called by the `wpcom` localization middleware, and that fails if the `i18n-calypso` library is an incomplete mock that doesn't define these methods.

**How to test:**
1. Set your locale to a non-English language and verify that all WP.com REST API requests have a `locale` or `_locale` query param with the current locale.
2. Check also the special cases for `sr_latin` and `de_formal` locales. These should load the `sr_latin` and `de_formal` translation files, send `language=sr_latin` and `language=de_formal` slugs with REST request, **but**: the `document.lang` in DOM should be set to `sr` and `de` respectively, and `sr` and `de` moment.js localizations should be loaded. These two locales are specialized ones that are unknown to services and libraries outside WP.com.

@Automattic/i18n: please verify that I understand the "locale variant" concept correctly and that it really applies only to `sr_latin` and `de_formal`.

### CODE REVIEW

File: packages/i18n-calypso/src/index.js [hunk 1]
```diff
@@ -6,16 +6,24 @@ import localizeFactory from './localize';
...
+// These should be deprecated eventually, exposing only the default `i18n` instance
 export const moment = i18n.moment;
 export const numberFormat = i18n.numberFormat.bind( i18n );
 export const translate = i18n.translate.bind( i18n );
 export const configure = i18n.configure.bind( i18n );
 export const setLocale = i18n.setLocale.bind( i18n );
 export const getLocale = i18n.getLocale.bind( i18n );
 export const getLocaleSlug = i18n.getLocaleSlug.bind( i18n );
+export const getLocaleVariant = i18n.getLocaleVariant.bind( i18n );
+export const isRtl = i18n.isRtl.bind( i18n );
```
Comment (sgomes) [inline packages/i18n-calypso/src/index.js]: Bad merge? This change looks unrelated.

**Review by sgomes — COMMENTED** [review]
Generally LGTM, code-wise. Left a small comment.

It would be best for someone with a stronger background in i18n than me to ensure correctness here.

File: packages/i18n-calypso/src/index.js (hunk 1 above)
Comment (jsnajdr) [reply packages/i18n-calypso/src/index.js]: That's a drive-by change that follows up on #37593 which added the `isRtl` method to the `I18N` class, but I forgot to export the bound version (for consistency's sake)

**Comment by jsnajdr** [comment]
Rebased on top of latest `master` -- let's see if the red e2e tests recover. The PR is still ready for review and merge.

**Comment by jsnajdr** [comment]
Revived this old (almost a year!) PR that simplifies the code that adds `_locale` and `locale` query params to all REST requests, depending on the current locale.

**Review by tyxla — APPROVED** [review]
Looks and works great! Nice cleanup @jsnajdr!

Requests to `rest/v*` properly get a `?locale=LOCALE` and requests to the `wpcom/v2` get a `?_locale=LOCALE` just like they did before.

Also, requests with `de_formal` work correctly in my testing. 

It would be useful to get a 👍 from @Automattic/i18n just in case.

**Review by yuliyan — APPROVED** [review]
Tested and it works fine - LGTM!

> @Automattic/i18n: please verify that I understand the "locale variant" concept correctly and that it really applies only to sr_latin and de_formal.

At the time of creating the PR, `sr_latin` and `de_formal` were indeed the only locale variants that we had, but currently we have two more `de-ch` and `es-mx`. I've tested with those and everything seems to be working fine there as well.

File: client/my-sites/plan-features/test/index.jsx [hunk 2]
```diff
@@ -7,19 +7,6 @@ jest.mock( 'lib/analytics/page-view', () => ( {} ) );
 jest.mock( 'lib/analytics/ad-tracking', () => ( {} ) );
 jest.mock( 'lib/analytics/page-view-tracker', () => 'PageViewTracker' );
 
-jest.mock( 'i18n-calypso', () => ( {
```
[the remark this answers is not in the document]
Comment (jsnajdr) [reply client/my-sites/plan-features/test/index.jsx:10]: Yes, there is a paragraph about why in the PR description.

First, mocking `i18n-calypso` is not needed to make the tests work. Maybe it was needed in the past when using the real `i18n-calypso` module needed some pre-configuration or dragged in too many other dependencies. But it's no longer needed today.

Second, the mock was incomplete, missing some methods that are actually called by the tests now after this patch.
